### PR TITLE
Store the Auth state in redux and inject auth status into components

### PR DESCRIPTION
**Pull request by nazar**
Refs #12 - various

1. Add Redux to the client to manage the auth0 authorisation state and be able to inject it into any component
2. Update checkJwt API middleware to passthrough any requests that do not contain an `authorisation` header
3. Split the client Auth service class into two modules: `login.js` to manage auth0 Lock and `webAuth` to manage auth0-js
4. tidy client component folder by renaming each component's entry file to `index.js` - this allows us to reference the folder when using imports - i.e. `import Callback from '../Callback'` vs `import Callback from '../Callback/Callback'`
5. removed CRA artifacts such as the react logo etc...

### CODE REVIEW

File: client/src/actions/auth.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import {authLogout} from '../services/login';
+
+export const AUTH_FAILURE = 'auth_failure';
```
Comment (mikelax) [inline client/src/actions/auth.js:3]: @nazar  should these const values be snake case or camelCase?
ie. `export const AUTH_FAILURE = 'authFailure';`

File: client/src/components/Callback/index.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import { Redirect } from 'react-router-dom';
+
+function Callback(props) {
+  const {isAuthenticated} = props.authorisation;
+
+  if (isAuthenticated) {
+    return <Redirect to="/"/>;
+  } else {
+    return <div className="text-center">
+      Authorising
```
Comment (mikelax) [inline client/src/components/Callback/index.js:12]: s -> z

haha, its all over. not a big deal until we get newer design updates.

File: client/src/components/Messages/Create/TestPing.js [hunk 3]
```diff
@@ -1,41 +0,0 @@
...
-export default class TestPing extends Component<{||}> {
-  constructor(props) {
-    super(props);
-
-    this.state = {};
-  }
-
-  securedPing = () => {
-    const { getAccessToken } = this.props.auth;
-    const API_URL = 'http://localhost:3000/api';
```
Comment (mikelax) [inline client/src/components/Messages/Create/TestPing.js:15]: @nazar  obviously this was always test code just for proof of concept in calling the api. 
Do you have an example of where the client is making an API call using "correct" pattern of constants for the URL etc?

**Review by mikelax — COMMENTED** [review]
@nazar one additional question related to `AuthGuard.js`
I see how that is a HOC that wraps a given component that needs an Auth otherwise it redirects to /login

I didn't see though where the specific check for scopes are done for each component? ie. create game will use different scope than create character or List Games for example?

**Comment by mikelax** [comment]
looking good @nazar . this is really nice! 
can you add a bit more info to the docs about some of the functions that are "redux" related I think. 
Things like the `mapStateToProps` or other redux related idioms?

File: client/src/services/webAuth.js [hunk 4]
```diff
@@ -45,6 +45,17 @@ export function clearRenewalTimer() {
...
 }
 
+export function userHasScopes(scopes) {
+  if (!localStorage.getItem('scopes')) {
+    return false;
+  }
+
+  const grantedScopes = JSON.parse(localStorage.getItem('scopes')).split(' ');
+
+  return scopes.every(scope => grantedScopes.includes(scope));
```
Comment (mikelax) [inline client/src/services/webAuth.js:55]: @nazar  is it correct to use the the `every` function here? Isn't the logic that the user must have at least one of the required scopes? Isn't that what we did for DS?

As an aside is this function `userHasScopes` referenced anywhere? I did a search of code and nothing came up?

File: client/src/actions/auth.js (hunk 1 above)
Comment (nazar) [reply client/src/actions/auth.js:3]: It doesn't matter what the string value is as it is never directly accessed - only the constant is.

File: client/src/services/webAuth.js (hunk 4 above)
Comment (nazar) [reply client/src/services/webAuth.js:55]: `userHasScopes` isn't referenced anywhere; I copied it from the Auth class for "safe keeping" until we need it :)

File: client/src/components/Callback/index.js (hunk 2 above)
Comment (nazar) [reply client/src/components/Callback/index.js:12]: hehe - I was planning on adding a big spinner (FontAwesome spinner came to mind) - but I didn't want to pull in a font icon library just yet

**Comment by nazar** [comment]
>I didn't see though where the specific check for scopes are done for each component? ie. create game will use different scope than create character or List Games for example?

Yep - I haven't added scope checks yet

**Comment by mikelax** [comment]
@nazar  thanks for the README info on redux, thunk, etc. very helpful. 🚀

**Review by mikelax — APPROVED** [review]

**Comment by nazar** [comment]
oh 🤦‍♂️ - I just realised there is a typo in `authorisation` -> `authorization` 🤦‍♂️🤦‍♂️🤦‍♂️

**Pull request merged**